### ISSUE TITLE
unwrapping conversion design space

### ISSUE THREAD

**Issue by dhollman** [executors/futures#13]
I don't know how much time we have to get into this, but we are going to *at least* have to specify a return type for `.then()`, so there will need to be some discussion along these lines. 
 `std::experimental::future` has an unwrapping constructor, but the design space for `future` unwrapping is much broader than this.  For instance, when a task produces subsequent hooks for different data that may attach to different continuations, about the only way to make that pattern work without promises is to return a `tuple<future<T1>,future<T2>>`. That means the return type of the corresponding `.then()` is `future<tuple<future<T1>,future<T2>>>`. Structured bindings can make this much more convenient to use if this type were to unwrap into a `tuple<future<T1>,future<T2>>`. For instance:

```c++
auto [fa, fb] = make_ready_future().then([]{
  auto fa = make_ready_future().then([]{
    // ... do something ...
    return 42;
  });
  auto fb = make_ready_future().then([]{
    // ... do something ...
    return 3.14;
  });
  return std::make_tuple(std::move(fa), std::move(fb));
});

fa.then([](auto a){ /*...*/ });
fb.then([](auto b){ /*...*/ });
```

The same sort of logic could be applied to mixed immediate/`future` return types:

```c++
auto [fa, fb] = make_ready_future().then([]{
  auto fb = make_ready_future().then([]{
    // ... do something ...
    return 3.14;
  });
  return std::make_tuple(42, std::move(fb));
});

fa.then([](auto a){ /*...*/ });
fb.then([](auto b){ /*...*/ });
```

Other design considerations related to unwrapping include:

- Using return type directly versus conversion: a converting constructor is not sufficient to write code with `auto` like most of us do these days. You could also just use a different name or something; I suspect the case where someone actually wants a `future<future<T>>` is niche enough that they'd be willing to do some extra typing.
- Exception propagation.
- Flattening: Does a `future<tuple<future<T>, future<tuple<future<U>, future<V>>>>` unwrap into a `tuple<future<T>, future<U>, future<V>>`? (Anyone who has ever tried to write `promise`-free arbitrary/non-strict fork-join code without unwrapping will tell you that this disaster of a type *does* actually come up)
- Generic code: does unwrapping `future<tuple<future<T>, future<U>>>` into a type with a different interface (i.e., a `tuple`) depending on the return type of the continuation closure make generic code too difficult? (Perhaps this is a reason to give the `tuple`-unwrapping form a different name?)
- Anything I'm missing here?

**Comment by brycelelbach** [comment]
Yah, this is something Michal and I talked about in P0701.

I don't think we want the future concepts to do fancy unwrapping with tuple, etc - that would add a lot of implementation burden on future authors. We may want the concrete future wrapper types to do this.

Marking Not Version 1 as it's additive and not absolutely and unequivocally fundamental.

**Comment by brycelelbach** [comment]
Also see #29 which addresses just the design space of the unwrapping constructor.

**Comment by dhollman** [comment]
> We may want the concrete future wrapper types to do this.

Agreed. That's the direction I would go for something like this — we definitely don't want it in the concept directly.  Just wanted to put it on the radar.

**Comment by brycelelbach** [comment]
This issue is hereby restricted to discussion of future unwrapping other than the unwrapping constructor, which is being discussed in #29. E.g. talk about unwrapping of `tuple`s and `variant`s for `when_*` interfaces.

**Comment by griwes** [comment]
My knee-jerk reaction is that I dislike this, because this way you encode the behavior for *all* instances of `future_t<T>`, while what I'd prefer to have is to be able to specify the behavior per call site, rather than per type. Sometimes I'll want to expand the tuple into the arguments... but sometimes I will definitely want to keep it unexpanded. If someone specializes the trait for tuples, I - as the end user of the API - no longer have the fine grained control that I want.

Going slightly off the rails here, this is why I'm not a fan of how Haskell typeclass instances work (but at the same time, why I'm not a fan of how template specializations work for customization points: you fix the behavior once, and then you can no longer have any other specialization without hard aliases. Prime example: Haskell's `Monoid`, integers, and `+` vs `*` that require you to `newtype` at least one of them to have both instances.

**Comment by griwes** [comment]
(I'm currently experimenting with designing and implementing a language that has typeclasses/traits/call it as you wish, and in addition to the "default" instances, you can define any number of other instances, that you have to explicitly name when you want to use them. If we had that here - i.e. some sensible behavior by default, but a way to override it - I'd be in favor.

I guess it should be possible to invent such a scheme in C++ - possibly by having an overload with an additional template parameter where I can override the trait by specifying a specific type to treat as the trait?)

**Comment by RedBeard0531** [comment]
@griwes 
> I'd prefer to have is to be able to specify the behavior per call site

Good point. That can be done by making the invocation traits be a defaulted template argument to `then()` that callers can pass explicitly when they want to override the default. I've updated the [gist](https://gist.github.com/RedBeard0531/f2466679b3acd06cd79d06e5cf29fda6/cd909473d1cef473fec259d472828a451a03e26a#file-future_traits-cpp-L137-L141) to allow that.

I think it only makes sense to provide control over how arguments are passed since the callback is already in control of its return type so it can always `return keep{dont_touch_this};`.

Does that address your concerns?

**Comment by griwes** [comment]
As my second comment indicated - yes, I'd be fine with something like that. As long as we also have a set of "standard" traits that do the obvious things ("never unwrap" and whatnot) in the standard at some point.

As this is an ABI break if we try to add it later - we'd need to either do this immediately, or have multiple overloads of `.then` (on the arity of the template parameter list). I'd prefer to do it immediately, but I don't know if we can get this specified sensibly in time for Rapperswil.

@brycelelbach @LeeHowes thoughts?

**Comment by RedBeard0531** [comment]
We already provide the `std::expected`/`Try`/`StatusWith<T>` unwrapping behavior in our codebase using a [more primitive](https://github.com/RedBeard0531/mongo_utils/blob/5d24c3dce16d99783255879fb72d8442003d57b8/util/future.h#L120-L239), less generic, mechanism for intercepting and adapting the return type from continuations. The feedback from our consumers is that it is extremely useful for simplifying their code.

Since `std::expected` doesn't exist yet, we can always change the concept to special case it later without technically being a breaking change since no code could be using a `Future<std::expected<T>>`. But we would need to signal intent early to ensure that Future implementers can prepare for that in a way that won't break ABI. Also, I'd prefer if we didn't special case a blessed stdlib type without a way for third_party libs to bless their own ErrorOr<T> types (does everyone have one at this point?). But I'd still prefer the special case to winding up with `future<std::expected<T>>` once both types are available.
